### PR TITLE
Refactor Git operations and auth

### LINKED ISSUE #1015: [FEATURE] Provide JGit `SshSessionFactory` via `setTransportConfigCallback`
**Is your feature request related to a problem? Please describe.**
We currently set a static `SshSessionFactory`, which opens us up to leaks.

**Describe the solution you'd like**
Use `setTransportConfigCallback` instead. 

**Describe alternatives you've considered**
Fine as it is now, but could be better.

**Additional context**
Blocked on #995.

### PR DESCRIPTION
**Pull request by fmeum**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## :scroll: Description

* Simplify GitOperation auth flow
* Simplify GitOperation inheritance structure

Note that the dialog shown in `BreakOutOfDetached`'s `onSuccess` may be finished immediately since it does not set `finishActivityOnEnd` to false. I didn't change anything about it, but it warrants a look.

## :bulb: Motivation and Context

Fixes #1015.

## :green_heart: How did you test it?
A bit, but this definitely needs more testing.

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [ ] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by msfjarvis** [comment]
What's here looks pretty good. While we're at it, I'd really prefer if we rip the bandaid off and also get rid of the `finishActivity*` stuff. I can try to get it done over the weekend after I finish fixing up the JGit Java7 compat layer and getting it to build with Gradle.

**Comment by fmeum** [comment]
> What's here looks pretty good. While we're at it, I'd really prefer if we rip the bandaid off and also get rid of the `finishActivity*` stuff. I can try to get it done over the weekend after I finish fixing up the JGit Java7 compat layer and getting it to build with Gradle.

:+1: :+1:

**Comment by msfjarvis** [comment]
I tested the following scenarios and noticed no regressions from `develop`:

- Deleting passwords
- Creating passwords
- Autofilling generated passwords

~~I think editing has regressed again with the race problem so I'll take a look at that.~~ Has been fixed

**Review by msfjarvis — DISMISSED** [review]
Approving this so Fabian can take a look and merge directly if it's good to go.

**Review by msfjarvis — DISMISSED** [review]
Turns out I can find mistakes better when I'm pissed

**Comment by fmeum** [comment]
I hope that I simplified `PasswordCreationActivity`, but these changes haven't been tested yet. They should help prevent double-committing in the future.

Also, I would really like to see us move the `onSuccess` overrides out of the `GitOperation`. That `SyncOperation` finishes the `callingActivity` there and `commitChange`'s `GitOperation` doesn't is, as I see it, purely a UI decision. Ideally, this would not be handled in the individual operations at all and would help us make changes here with more confidence (and less unintended finishing of activities). @msfjarvis Do you have an idea how to achieve this?

**Comment by msfjarvis** [comment]
> I hope that I simplified `PasswordCreationActivity`, but these changes haven't been tested yet. They should help prevent double-committing in the future.

I think it's gonna break during the pre-existing file checks but I'll try and confirm in a bit.

> Also, I would really like to see us move the `onSuccess` overrides out of the `GitOperation`. That `SyncOperation` finishes the `callingActivity` there and `commitChange`'s `GitOperation` doesn't is, as I see it, purely a UI decision. Ideally, this would not be handled in the individual operations at all and would help us make changes here with more confidence (and less unintended finishing of activities). @msfjarvis Do you have an idea how to achieve this?

One idea for that would be to make `GitCommandExecutor#execute` return either a `Result<T>` type or enum and we forward that value to the activities. I also *really* wanna get rid of `GitOperationActivity` and make each individual activity handles its git tasks.

**Comment by fmeum** [comment]
Feel free to revert, but I pushed a commit that switches us over to the Kotlin `Result` type.

I also noticed that `handleResult` is never called. I generally quite like the idea of having execute return a result though and letting the activities handle it themselves. The switch away from `GitOperationActivity` should probably be saved for another PR though.

**Comment by msfjarvis** [comment]
> Feel free to revert, but I pushed a commit that switches us over to the Kotlin `Result` type.

That's perfectly fine.

> I also noticed that `handleResult` is never called.

Right, that's because I hadn't yet settled on how to get results down to individual activities and whether or not I wanna have a "default" handler for activities to use. I've set time aside later this week to wrap up both this PR and #1068.

> I generally quite like the idea of having execute return a result though and letting the activities handle it themselves. The switch away from `GitOperationActivity` should probably be saved for another PR though.

I wanted to do it here because getting rid of it would open up more call-sites for us to verify that the changes we're making for GitOperation to pass down results are compatible for all our individual use-cases. I'm perfectly fine with delegating it to a future point in time as well.

**Comment by fmeum** [comment]
> > Feel free to revert, but I pushed a commit that switches us over to the Kotlin `Result` type.
> 
> That's perfectly fine.
> 
> > I also noticed that `handleResult` is never called.
> 
> Right, that's because I hadn't yet settled on how to get results down to individual activities and whether or not I wanna have a "default" handler for activities to use. I've set time aside later this week to wrap up both this PR and #1068.
> 
> > I generally quite like the idea of having execute return a result though and letting the activities handle it themselves. The switch away from `GitOperationActivity` should probably be saved for another PR though.
> 
> I wanted to do it here because getting rid of it would open up more call-sites for us to verify that the changes we're making for GitOperation to pass down results are compatible for all our individual use-cases. I'm perfectly fine with delegating it to a future point in time as well.

Totally fine, let's reclassify this as a draft though in that case. I just thought you wanted to get this in quickly in which case I wouldn't have felt comfortable making so many changes at once.

**Comment by msfjarvis** [comment]
Think this is about ready now. I've tested most cases but I'd appreciate some more eyes around tricky bits.

File: app/proguard-rules.pro
Comment (msfjarvis) [inline app/proguard-rules.pro:21]: I was getting errors in the log about not being able to resolve the Random class from jsch, not sure if that's mission-critical so feel free to drop this if it's not going to break things.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt]: This now also refreshes during push where its not strictly needed but it simplified the code so I let it be.

File: app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt]: I added these to simplify general cases for all callers to mirror the now defunct `onSuccess`/`onError` API from `GitOperation`. It unfortunately now needs to be called manually by each caller but I don't have any better ideas.

File: app/src/main/java/com/zeapo/pwdstore/git/GitConfigActivity.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/git/GitConfigActivity.kt:88]: We're losing the exact branch name here but I think we'll be fine.

File: app/src/main/java/com/zeapo/pwdstore/git/operation/GitOperation.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/git/operation/GitOperation.kt]: I had to do this because otherwise we were finishing away `PasswordStore` on biometric cancellation.

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt:63]: This is safely nullable and fixes the Snackbar position in `PasswordFragment`

File: app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/utils/Extensions.kt:112]: We'll need to audit this change and confirm there's no callers left who relied on the implicit finish and haven't been fixed.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt]: Totally fine, refreshing is reasonably cheap.

File: app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt]: I also don't know how to improve that design and think it's okay now that it happens on the caller side.

How about we name these `finishOnSuccessHandler` and `finishAfterPromptOnErrorHandler` (or something similar) though?

File: app/src/main/java/com/zeapo/pwdstore/git/GitConfigActivity.kt
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/git/GitConfigActivity.kt:88]: I think so too

File: app/proguard-rules.pro
Comment (fmeum) [reply app/proguard-rules.pro:21]: Well, whatever. Let's leave it in and stop worrying about Jsch alltogether.

File: app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/git/BaseGitActivity.kt]: Sure I'll rename these in a bit.

File: app/src/main/java/com/zeapo/pwdstore/git/operation/GitOperation.kt
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/git/operation/GitOperation.kt]: It might be cleaner to throw an `SSHException` with `disconnectReason` `AUTH_CANCELLED_BY_USER` here, but not by much. Your call.

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/git/operation/GitOperation.kt]: I'll make the change.

**Comment by fmeum** [comment]
I'm observing a crash when I cancel an SSH password prompt. Maybe the `UserAuthException` gets swallowed somehwere? 

Edit: or rather isn't caught?

**Comment by msfjarvis** [comment]
> I'm observing a crash when I cancel an SSH password prompt. Maybe the `UserAuthException` gets swallowed somehwere? 
> 
> Edit: or rather isn't caught?

I'll take a look.

**Comment by msfjarvis** [comment]
Not *completely* sure yet why we're failing to catch our exceptions. I got some work stuff to do right now so I'll have to dive into this later today.

**Comment by msfjarvis** [comment]
Spent about an hour, didn't make any progress. Structured concurrency is hard :-)

**Comment by fmeum** [comment]
> Spent about an hour, didn't make any progress. Structured concurrency is hard :-)

I think I now know why this is happening, bit don't yet know how we should fix it. Will think about it.

**Comment by fmeum** [comment]
> > Spent about an hour, didn't make any progress. Structured concurrency is hard :-)
> 
> I think I now know why this is happening, bit don't yet know how we should fix it. Will think about it.

What I thought would be the cause of the issue turned out not to be it. Keep in mind though that `lifecycleScope` attaches a `SupervisorJob`, which based on my understanding of coroutines means that exceptions crossing the boundaries of a child coroutine will fail the parent, i.e. crash our app. It seems important to use `try...catch` in the coroutine that throws the exception.

I have bisected the issue and first observe the crash with 0581157aecc76426bafa9b213a82916f8454377a. I will see whether I can get to the bottom of this.

**Comment by msfjarvis** [comment]
> > > Spent about an hour, didn't make any progress. Structured concurrency is hard :-)
> > 
> > 
> > I think I now know why this is happening, bit don't yet know how we should fix it. Will think about it.
> 
> What I thought would be the cause of the issue turned out not to be it. Keep in mind though that `lifecycleScope` attaches a `SupervisorJob`, which based on my understanding of coroutines means that exceptions crossing the boundaries of a child coroutine will fail the parent, i.e. crash our app. It seems important to use `try...catch` in the coroutine that throws the exception.

That's what I thought as well. I tried wrapping the call to `launchGitOperation` in a nested scope to prevent it from being bubbled upwards to the parent. Maybe a new `supervisorScope` is a necessity, but then we can't properly extract the return value without making each call site of `launchGitOperation` roll its own `supervisorScope`.

**Comment by fmeum** [comment]
We learned something about experimental API usage the hard way here:

```kotlin
package com.zeapo.pwdstore

import java.lang.IllegalStateException
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.junit.Test

class ResultTest {

    @Test
    fun testCoroutineNativeResult() {
        runBlocking {
            launch {
                fooNative()
            }
        }

    }

    suspend fun fooNative(): Result<Unit> {
        delay(10)
        return Result.failure(IllegalStateException("should not be thrown"))
    }

    @Test
    fun testCoroutineCustomResult() {
        runBlocking {
            launch {
                fooCustom()
            }
        }
    }

    sealed class CustomResult<T> {
        class Failure<T>(t: Throwable) : CustomResult<T>()
        class Success<T>(value: T): CustomResult<T>()
    }

    suspend fun fooCustom(): CustomResult<Unit> {
        delay(10)
        return CustomResult.Failure(IllegalStateException("should not be thrown"))
    }
}
```

I recommend giving this a try and seeing for yourself ;-)

**Comment by msfjarvis** [comment]
I am now unfathomably mad and I blame you.

(I'm kidding)

**Comment by fmeum** [comment]
> I am now unfathomably mad and I blame you.

How about this result type instead: https://github.com/michaelbull/kotlin-result? I'm somewhat paranoid about stuff with compiler integration now...

**Comment by msfjarvis** [comment]
> > I am now unfathomably mad and I blame you.
> 
> How about this result type instead: [michaelbull/kotlin-result](https://github.com/michaelbull/kotlin-result)? I'm somewhat paranoid about stuff with compiler integration now...

I've wanted to try this for a bit, seems like a good time now. I'll get on it.

**Comment by msfjarvis** [comment]
And _obviously_, nothing crashes anymore...

**Review by msfjarvis — DISMISSED** [review]
Approving so you can merge it yourself.

**Comment by fmeum** [comment]
I cleaned up our existing mix of `try..catch` and `Result` in `GitCommandExecutor` and made it so that the FAB moves up for the snackbar, which afaik is the best practice for handling the two.

**Comment by fmeum** [comment]
Feel free to merge if this looks good to you, I haven't found any other issues.

**Comment by msfjarvis** [comment]
> I cleaned up our existing mix of `try..catch` and `Result` in `GitCommandExecutor` and made it so that the FAB moves up for the snackbar, which afaik is the best practice for handling the two.

A long time ago, in a galaxy far, far away, Google [decided that FAB moves for no-one](https://material.io/components/snackbars#placement).

**Comment by fmeum** [comment]
> > I cleaned up our existing mix of `try..catch` and `Result` in `GitCommandExecutor` and made it so that the FAB moves up for the snackbar, which afaik is the best practice for handling the two.
> 
> A long time ago, in a galaxy far, far away, Google [decided that FAB moves for no-one](https://material.io/components/snackbars#placement).

I sort of liked this fantasy galaxy I have been living in thus far, but I see that it's time to move on. Thanks for the link!

**Review by msfjarvis — APPROVED** [review]

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:84]: I just noticed that this commit broke the "Back" button folder navigation and I think it's because of `BaseGitActivity`'s `onOptionsItemSelected` override. Maybe it's time to get rid of it?

**Review by fmeum — COMMENTED** [review]
@msfjarvis  Found that this breaks folder navigation and have an idea of why.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:84]: Yeah. I'll put up a PR.

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:84]: Pushed #1075

**Pull request merged**
